snuba/query/parser/conditions.py

```diff
@@ -0,0 +1,225 @@
+            lit = tuple(sorted(lit))
+
+        # If the LHS is a simple column name that refers to an array column
+        # (and we are not arrayJoining on that column, which would make it
+        # scalar again) and the RHS is a scalar value, we assume that the user
+        # actually means to check if any (or all) items in the array match the
+        # predicate, so we return an `any(x == value for x in array_column)`
+        # type expression. We assume that operators looking for a specific value
+        # (IN, =, LIKE) are looking for rows where any array value matches, and
+        # exclusionary operators (NOT IN, NOT LIKE, !=) are looking for rows
+        # where all elements match (eg. all NOT LIKE 'foo').
+        columns = dataset.get_dataset_schemas().get_read_schema().get_columns()
+        if (
+            isinstance(lhs, str)
+            and lhs in columns
+            and isinstance(columns[lhs].type, Array)
+            and columns[lhs].base_name != array_join
+            and not isinstance(lit, (list, tuple))
+        ):
+            return unpack_array_condition_builder(operand_builder(lhs), op, lit)
+        else:
+            return simple_condition_builder(operand_builder(lhs), op, lit)
+
+    elif depth == 1:
+        sub = (
+            parse_conditions(
+                operand_builder,
+                and_builder,
+                or_builder,
+                unpack_array_condition_builder,
+                simple_condition_builder,
+                dataset,
+                cond,
+                array_join,
+                depth + 1,
+            )
+            for cond in conditions
+        )
+        return or_builder([s for s in sub if s])
+    else:
+        raise InvalidConditionException(str(conditions))
+
+
+def parse_conditions_to_expr(
+    expr: Sequence[Any], dataset: Dataset, arrayjoin: Optional[str]
+) -> Optional[Expression]:
+    """
+    Relies on parse_conditions to parse a list of conditions into an Expression.
+    """
+
+    def operand_builder(val: Any) -> Expression:
+        if is_function(val, 0):
+            return parse_function_to_expr(val)
+        # TODO: This will use the schema of the dataset to decide
+        # if the expression is a column or a literal.
+        if QUOTED_LITERAL_RE.match(val):
+            return Literal(None, val[1:-1])
+        else:
+            return Column(None, val, None)
+
+    def multi_expression_builder(
+        expressions: Sequence[Expression], function: str
+    ) -> Expression:
+        assert len(expressions) > 0
+        if len(expressions) == 1:
+            return expressions[0]
+
+        return binary_condition(
+            None,
+            function,
+            expressions[0],
+            multi_expression_builder(expressions[1:], function),
+        )
+
+    def and_builder(expressions: Sequence[Expression]) -> Optional[Expression]:
+        if not expressions:
+            return None
```

Comment:
This is for the type checker to keep the return type of multi_expression_builder non nullable which helps with the recursion.